trax/data/tf_inputs.py

```diff
@@ -1107,6 +1112,42 @@ def bert_double_sentence_preprocess(batch):
     yield value_vector, segment_embs, segment_embs, label, np.int64(1)
 
 
+def mask_random_tokens(batch, vocab_size=30522, masking_prob=0.15, merge_with_loss=True):
+  """
+  Follows standard procedure of masking in BERT:
+  `masking_prob` random tokens are selected out of which:
+    80% are replaced with [MASK] token
+    10% are replaced with random token
+    10% are leaved with original token unchanged
+
+  ids of the tokens are from original lowercase vocab
+  """
+
+  for token_ids, *row_rest in batch:
+    y = token_ids.copy()
+
+    is_special_token = np.logical_or(token_ids == 101, token_ids == 102) # CLS and SEP tokens
+    is_special_token = np.logical_or(is_special_token, token_ids == 0) # padding
+    prob_scores = np.maximum(np.random.random(token_ids.shape), is_special_token)
+    tokens_to_predict = prob_scores < masking_prob # loss will be calculated only on these tokens
+    total_tokens_to_pred = tokens_to_predict.sum()
+    if total_tokens_to_pred > 0:
+      token_weights = tokens_to_predict.astype(np.float32) / total_tokens_to_pred # total sum is 1
+
+    # , 10% to random token and leave 10% unchanged
+    mask_threshold, random_token_threshold = 0.8 * masking_prob, 0.9 * masking_prob
+    # change 80 % of tokens to [MASK]
+    mask_tokens = prob_scores < mask_threshold
+    token_ids[mask_tokens] = 103
+
+    # change 10% of tokens to random token
+    random_tokens = np.logical_and(prob_scores > mask_threshold, prob_scores < random_token_threshold)
+    token_ids[random_tokens] = np.random.randint(999, vocab_size, random_tokens.sum())
```

Comment:
Can you explain a little bit in comments what's going on here, why is 999 special?

Reply:
999 is the first token id in standard bert_base_lowercase vocab. I will add this to arguments just like other special ids.

Here for 10% of tokens selected for prediction are changed to random tokens.
This is done in this way because of the sampling methodology i used. Instead of using 2 random samples: 
1) First whether given will be used to evaluation (0.15 probability in standard version)
2) split selected 15% of tokens into 3 groups: (0.8 MASK, 0.1 random token, 0.1 leave original)
Which is the way it is described.
I do one split which splits it into 4 groups: (p_choose is 0.15 in original version)
  (0.8 * p_choose) MASK, 
  (0.1 * p_choose) random token, 
  (0.1 * p_choose) original tokens, 
  (1 - p_choose) tokens not predicted.
Which uses just one sampling and is the same experiment. 
I'll add some comment to the docstring about it for clarity

Reply:
I checked out the original bert code and realized that in original bert 15% of tokens are always chosen (instead of choosing for each token independently with 15% probability whether it is involved in masking). So I rewritten the code for consistency with original method. I also extended the documentation quite significantly